Sources/BraintreeThreeDSecure/BTThreeDSecureRequest.swift

```diff
@@ -0,0 +1,377 @@
+                }
+            }
+        }
+    }
+    
+    private func performV2Authentication(with lookupResult: BTThreeDSecureResult) {
+        guard let apiClient = paymentFlowClientDelegate?.apiClient() else {
+            paymentFlowClientDelegate?.onPaymentComplete(nil, error: BTThreeDSecureError.invalidAPIClient)
+            return
+        }
+
+        threeDSecureV2Provider?.process(lookupResult: lookupResult) { result, error in
+            guard let result else {
+                apiClient.sendAnalyticsEvent("ios.three-d-secure.verification-flow.failed")
+                self.paymentFlowClientDelegate?.onPaymentComplete(nil, error: error)
+                return
+            }
+
+            self.logThreeDSecureCompletedAnalytics(forResult: lookupResult, apiClient: apiClient)
+            self.paymentFlowClientDelegate?.onPaymentComplete(result, error: error)
+        }
+    }
+    
+    private func logThreeDSecureCompletedAnalytics(forResult result: BTThreeDSecureResult, apiClient: BTAPIClient) {
+        let liabilityShiftPossible = result.tokenizedCard?.threeDSecureInfo.liabilityShiftPossible ?? false
+        apiClient.sendAnalyticsEvent("ios.three-d-secure.verification-flow.liability-shift-possible.\(stringFor(liabilityShiftPossible))")
+
+        let liabilityShifted = result.tokenizedCard?.threeDSecureInfo.liabilityShiftPossible ?? false
+        apiClient.sendAnalyticsEvent("ios.three-d-secure.verification-flow.liability-shifted.\(liabilityShifted)")
+
+        apiClient.sendAnalyticsEvent("ios.three-d-secure.verification-flow.completed")
+    }
+    
+    func stringFor(_ boolean: Bool) -> String {
+        boolean ? "true" : "false"
+    }
+}
+
+// MARK: - BTPaymentFlowRequestDelegate Protocol Conformance
+
+extension BTThreeDSecureRequest: BTPaymentFlowRequestDelegate {
+
+    /// :nodoc:
+    public func handle(
+        _ request: BTPaymentFlowRequest,
+        client apiClient: BTAPIClient,
+        paymentClientDelegate delegate: BTPaymentFlowClientDelegate
+    ) {
+        paymentFlowClientDelegate = delegate
+
+        apiClient.sendAnalyticsEvent("ios.three-d-secure.initialized")
+
+        apiClient.fetchOrReturnRemoteConfiguration { [weak self] configuration, error in
+            guard let self else { return }
+
+            if let error {
+                self.paymentFlowClientDelegate?.onPaymentComplete(nil, error: error)
+                return
+            }
+
+            var integrationError: Error?
+
+            if configuration?.cardinalAuthenticationJWT == nil {
+                NSLog("%@ BTThreeDSecureRequest versionRequested is 2, but merchant account is not setup properly.", BTLogLevelDescription.string(for: .critical))
+                integrationError = BTThreeDSecureError.configuration("BTThreeDSecureRequest versionRequested is 2, but merchant account is not setup properly.")
+            }
+
+            if self.amount?.decimalValue.isNaN == true {
```

Comment:
I'm still thinking about case where amount is nil.

Reply:
I c in Playground that this doesn't return true if amount is nil. It returns a nil.

Reply:
That's odd that it returns nil but I guess makes sense if amount is optional. Added an additional or nil check along with the `isNaN` check here: https://github.com/braintree/braintree_ios/pull/996/commits/80a883725a3c3e69c0ea8f8ec8018040b8f538fa

Reply:
I think it's weird too.